Source/HTTP/Curl/CurlProvider.cpp

```diff
@@ -133,28 +141,26 @@ HRESULT CALLBACK CurlProvider::CleanupAsyncProvider(XAsyncOp op, const XAsyncPro
         size_t multiIndex{ 0 };
         bool cleanupComplete{ false };
 
-        for (auto& pair : provider->m_curlMultis)
-        {
-            HRESULT hr = CurlMulti::CleanupAsync(std::move(pair.second), &provider->m_multiCleanupAsyncBlocks[multiIndex++]);
-            if (FAILED(hr))
-            {
-                // Continue cleanup if this fails, but we should expect 1 fewer MultiCleanupComplete callback
-                HC_TRACE_ERROR_HR(HTTPCLIENT, hr, "CurlMulti::CleanupAsync failed, continuing cleanup");
-
-                std::lock_guard<std::mutex> lock{ provider->m_mutex };
-                --provider->m_cleanupTasksRemaining;
-            }
-        }
-
-
-        {
-            std::lock_guard<std::mutex> lock{ provider->m_mutex };
-            if (--provider->m_cleanupTasksRemaining == 0)
-            {
-                // If there are no pending pending multi cleanups, complete cleanup here
-                cleanupComplete = true;
-            }
-        }
+		{
+			std::lock_guard<std::mutex> lock{ provider->m_mutex };
+			for (auto& pair : provider->m_curlMultis)
+			{
+				HRESULT hr = CurlMulti::CleanupAsync(std::move(pair.second), &provider->m_multiCleanupAsyncBlocks[multiIndex++]);
```

Comment:
>HRESULT hr = CurlMulti::CleanupAsync [](http://example.com/codeflow?start=4&length=36)

I believe it is not safe to call CurlMulti::CleanupAsync while holding m_mutex here.  CleanupAsync can complete synchronously in same cases, calling back into CurlProvider::MultiCleanupComplete which will try to grab the mutex.

CurlProvider::CleanupAsyncProvider has started, m_curlMultis should never be modified by another thread again - it should be safe to not lock here.

That said, to be safe we can grab the mutex, move m_curlMultis into a local, then release the lock and continue with cleanup as before but instead operating on the local curlMultis list

Reply:
Okay, I've put up a revised version.  Thanks.